lib/demo_web/controllers/user_auth.ex

```diff
@@ -0,0 +1,139 @@
+defmodule DemoWeb.UserAuth do
+  import Plug.Conn
+  import Phoenix.Controller
+
+  alias Demo.Accounts
+  alias DemoWeb.Router.Helpers, as: Routes
+
+  # Make the remember me cookie valid for 60 days.
+  # If you want bump or reduce this value, also change
+  # the token expiry itself in UserToken.
+  @max_age 60 * 60 * 24 * 60
+  @remember_me_cookie "user_remember_me"
+  @remember_me_options [sign: true, max_age: @max_age]
+
+  @doc """
+  Logs the user in.
+
+  It renews the session ID and clears the whole session
+  to avoid fixation attacks. See the renew_session
+  function to customize this behaviour.
+  """
+  def login_user(conn, user, params \\ %{}) do
+    token = Accounts.generate_session_token(user)
+    user_return_to = get_session(conn, :user_return_to)
+
+    conn
+    |> renew_session()
+    |> put_session(:user_token, token)
+    |> maybe_write_remember_me_cookie(token, params)
+    |> redirect(to: user_return_to || signed_in_path(conn))
+  end
+
+  defp maybe_write_remember_me_cookie(conn, token, %{"remember_me" => "true"}) do
+    put_resp_cookie(conn, @remember_me_cookie, token, @remember_me_options)
+  end
+
+  defp maybe_write_remember_me_cookie(conn, _token, _params) do
+    conn
+  end
+
+  # This function renews the session ID and erases the whole
+  # session to avoid fixation attacks. If there is any data
+  # in the session you may want to preserve after login/logout,
+  # you must explicitly fetch the session data before clearing
+  # and then immediately set it after clearing, for example:
+  #
+  #     defp renew_session(conn) do
+  #       preferred_locale = get_session(conn, :preferred_locale)
+  #
+  #       conn
+  #       |> configure_session(renew: true)
+  #       |> clear_session()
+  #       |> put_session(:preferred_locale, preferred_locale)
+  #     end
+  #
+  defp renew_session(conn) do
+    conn
+    |> configure_session(renew: true)
+    |> clear_session()
+  end
+
+  @doc """
+  Logs the user out.
+
+  It clears all session data for safety. See renew_session.
+  """
+  def logout_user(conn) do
+    user_token = get_session(conn, :user_token)
+    user_token && Accounts.delete_session_token(user_token)
+
+    conn
+    |> renew_session()
+    |> delete_resp_cookie(@remember_me_cookie)
+    |> redirect(to: "/")
+  end
+
+  @doc """
+  Authenticates the user by looking into the session
+  and remember me token.
+  """
+  def fetch_current_user(conn, _opts) do
```

Comment:
Would it be an idea to check if a current user already exist? 

```ex
def resolve_current_user(conn = %{assigns: %{current_user: user}}, _opts)
    when not is_nil(user) do
  conn
end

def resolve_current_user(conn, _opts) do
  {user_token, conn} = ensure_user_token(conn)
  user = user_token && Accounts.get_user_by_session_token(user_token)
  assign(conn, :current_user, user)
end
```

This way you can make "authenticated" tests way easier through tags;

```ex
setup tags do
  :ok = Sandbox.checkout(ExampleWeb.Repo)

  unless tags[:async] do
    Sandbox.mode(Tijdwerk.Repo, {:shared, self()})
  end

  user =
    cond do
      tags[:as_user] -> Tijdwerk.Factory.build(:user)
      tags[:as_inserted_user] -> Tijdwerk.Factory.insert(:user)
    true -> nil
  end

  conn =
    Phoenix.ConnTest.build_conn()
    |> Plug.Conn.assign(:current_user, user)

  {:ok, conn: conn, user: user}
end
```

Reply:
Your call! This would be necessary in the past but setting a test session is also very straight-forward on latest Phoenix.